src/main/java/net/minecraftforge/client/event/RegisterSpriteSourceTypesEvent.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.client.event;
+
+import com.google.common.collect.BiMap;
+import com.mojang.serialization.Codec;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.texture.atlas.SpriteSource;
+import net.minecraft.client.renderer.texture.atlas.SpriteSourceType;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.fml.LogicalSide;
+import net.minecraftforge.fml.event.IModBusEvent;
+import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
+
+/**
+ * Fired to allow mods to register their own SpriteSourceTypes.
+ * This event is fired once during the construction of the {@link Minecraft} instance.
+ *
+ * <p>This event is not {@linkplain Cancelable cancellable}, and does not {@linkplain HasResult have a result}.</p>
+ *
+ * <p>This event is fired on the {@linkplain FMLJavaModLoadingContext#getModEventBus() mod-specific event bus},
+ * only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterSpriteSourceTypesEvent extends Event implements IModBusEvent
+{
+    private final BiMap<ResourceLocation, SpriteSourceType> types;
+
+    public RegisterSpriteSourceTypesEvent(BiMap<ResourceLocation, SpriteSourceType> types)
+    {
+        this.types = types;
+    }
+
+    /**
+     * Registers the given {@link Codec} as SpriteSourceType under the given id.
+     *
+     * @param id    The id to register the {@link SpriteSourceType} under
+     * @param codec The codec for the {@link SpriteSourceType} to register
+     */
+    public void register(ResourceLocation id, Codec<? extends SpriteSource> codec)
+    {
+        this.types.put(id, new SpriteSourceType(codec));
```

Comment:
This should have the same validation as `SpriteSources#register`. You should not be able to override a `SpriteSourceType` entry.